### PR TITLE
Add FakeTensorMode

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78524
* #78523
* #78522
* #78517
* #78270
* #78090
* #78516
* __->__ #77972
* #77971
* #77970
* #77969

This adds a mode which will intercept calls to `__torch__dispatch__` even if the inputs are not already `FakeTensors`. This mimics the convenient [prior existing usage](https://pytorch.org/torchdistx/latest/fake_tensor.html). It does so by wrapping input tensors to Fake Tensors and then continuing to run the operators. 

Not Yet Implemented:

I still need to memoize conversion of non-fake tensors to fake tensors (and internally, to `meta` devices) following along with the [class here](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L70). 

One open question is what should be the duration of the `FakeTensorConverter`. IMO, it would make sense & be convenient for it to live for the duration of `FakeTensorMode`. Since we shouldn't be allocating any new Tensors with actual data (just on `meta` devices) it is probably fine for those tensors to live for the duration of `FakeTensorMode`.

If that is not sufficient, we could try using `weakref.WeakKeyDictionary` mapping tensors to their fake equivalents. I looked into this a bit and there are at least some a few incompatibilities that need to be dealt with.

### CODE REVIEW

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -156,3 +197,6 @@ def merge_devices(t):
         assert common_device is not None, f"Could not find common device for {func}"
 
         return common_device
+
+class FakeTensorMode(FakeTensor):
+    context = no_dispatch
```
Comment (ezyang) [inline torch/_subclasses/fake_tensor.py:202]: Can we do a modern style mode instead pretty please :)

File: torch/_subclasses/fake_tensor.py [hunk 2]
```diff
@@ -19,6 +21,29 @@
...
+@functools.lru_cache(None)
+def _is_tensor_constructor(func: OpOverload):
+    assert isinstance(func, OpOverload)
+    schema = func._schema
+    if any(contains_tensor_types(arg.type) for arg in schema.arguments):
+        return False
+    # TODO: no real reason to restrict multiple outputs
+    return (
+        len(schema.returns) == 1 and schema.returns[0].type is torch._C.TensorType.get()
+    )
```
Comment (ezyang) [inline torch/_subclasses/fake_tensor.py:44]: Tag opportunity :) cc @anjali411

File: torch/_subclasses/fake_tensor.py [hunk 3]
```diff
@@ -70,22 +108,25 @@ def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
...
+                )
 
-        r = super().__torch_dispatch__(func, types, args, kwargs)
+        if _is_tensor_constructor(func):
+            assert not func in _non_kwarg_device_constructors
+            _, new_kwargs = normalize_function(
+                func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
+            )
+            # cpu is default device if none is specified
+            out_device = new_kwargs.pop("device", torch.device("cpu"))
```
Comment (ezyang) [inline torch/_subclasses/fake_tensor.py:121]: technically it's `torch.get_default_tensor_type()` lol but ok

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:202]: You didn't actually use context, AFAICT?

**Review by ezyang — APPROVED** [review]

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (eellison) [reply torch/_subclasses/fake_tensor.py:202]: What is a modern style mode ? I believe the `context` stopped recursion

File: torch/_subclasses/fake_tensor.py (hunk 3 above)
Comment (eellison) [reply torch/_subclasses/fake_tensor.py:121]: which is immutable 😛

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py]: inherit from TorchDispatchMode

Comment (eellison) [reply torch/_subclasses/fake_tensor.py]: will do.. mind linking me differences between two why one should do modern style over existing ?

Comment (ezyang) [reply torch/_subclasses/fake_tensor.py]: the main difference is you can store instance variables on the mode, since it is an actual object not a class

Comment (eellison) [reply torch/_subclasses/fake_tensor.py]: nice... i probably don't need `setup_mode` then

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36618464).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36618464).

**Comment by eellison** [comment]
I actually think the current behavior makes sense. The easiest mental model IMO is that everything is simulated, and that no real tensors will be affected. I think allowing inplacing views and other things to actually affect the input tensors would be a mistake. 
If you do you would get into a situation where 
`input.t_()` does affect your input, but `input.add_(fake_tensor)` doesn't...

Additionally, the behavior above with `resize_` should just work because the first use of that tensor will get converted to fake, and then any subsequent uses the cached version with the `resize_` applied will be used, which correctly simulates what compute you would have done.

File: torch/_subclasses/fake_tensor.py [hunk 4]
```diff
@@ -20,6 +22,29 @@
...
+    return type.isSubtypeOf(tensor_type) or any(
+        contains_tensor_types(e) for e in type.containedTypes()
+    )
+
+
+@functools.lru_cache(None)
+def _is_tensor_constructor(func: OpOverload):
+    assert isinstance(func, OpOverload)
+    schema = func._schema
+    if any(contains_tensor_types(arg.type) for arg in schema.arguments):
```
Comment (anjali411) [inline torch/_subclasses/fake_tensor.py:40]: what about `_like` ops?

Comment (eellison) [reply torch/_subclasses/fake_tensor.py:40]: those got added later

**Pull request closed without merging**